File: experimental/sso-with-skills/csharp_dotnetcore/RootBot/Dialogs/SsoSignInDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Schema;

namespace Microsoft.BotBuilderSamples.RootBot.Dialogs
{
    public class SsoSignInDialog : ComponentDialog
    {
        public SsoSignInDialog(string connectionName)
            : base(nameof(SsoSignInDialog))
        {
            AddDialog(new OAuthPrompt(nameof(OAuthPrompt), new OAuthPromptSettings
            {
                ConnectionName = connectionName,
                Text = "Sign in to the host bot using AAD for SSO",
                Title = "Sign In"
            }));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { SignInStepAsync, DisplayTokenAsync }));
            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> SignInStepAsync(WaterfallStepContext context, CancellationToken cancellationToken)
        {
            return await context.BeginDialogAsync(nameof(OAuthPrompt), null, cancellationToken);
        }

        private async Task<DialogTurnResult> DisplayTokenAsync(WaterfallStepContext context, CancellationToken cancellationToken)
        {
            if (!(context.Result is TokenResponse result))
            {
                await context.Context.SendActivityAsync("No token was provided.", cancellationToken: cancellationToken);
            }
            else
            {
                await context.Context.SendActivityAsync($"Here is your token: {result.Token}", cancellationToken: cancellationToken);
            }

            return await context.EndDialogAsync(null, cancellationToken);
        }
    }
}

File: experimental/sso-with-skills/javascript_nodejs/rootBot/dialogs/ssoSignInDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ComponentDialog, OAuthPrompt, WaterfallDialog } = require('botbuilder-dialogs');

const WATERFALL_DIALOG = 'waterfallDialog';
const SSO_SIGN_IN_DIALOG = 'signInDialog';
const OAUTH_PROMPT = 'oAuthPrompt';

class SsoSignInDialog extends ComponentDialog {
    constructor(connectionName) {
        super(SSO_SIGN_IN_DIALOG);

        this.addDialog(new OAuthPrompt(OAUTH_PROMPT, {
            connectionName,
            text: 'Sign in to the root bot using Azure AD for SSO',
            title: 'Sign In'
        }))
            .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
                this.signInStep.bind(this),
                this.displayTokenStep.bind(this)
            ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    /**
     * Display the OAuthPrompt, asking the user to log in.
     */
    async signInStep(stepContext) {
        return stepContext.beginDialog(OAUTH_PROMPT);
    }

    /**
     * Display the user's token and end the dialog.
     */
    async displayTokenStep(stepContext) {
        if (!stepContext.result.token) {
            await stepContext.context.sendActivity('No token was provided.');
        } else {
            await stepContext.context.sendActivity(`Here is your token: ${ stepContext.result.token }`);
        }

        return stepContext.endDialog();
    }
}

module.exports = {
    SsoSignInDialog,
    SSO_SIGN_IN_DIALOG
};
